lib/srv/transport/stream.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transport
+
+import (
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/gravitational/trace"
+)
+
+// Stream abstracts gRPC streams used to transport
+// frames of data.
+type Stream interface {
+	Receive() ([]byte, error)
+	Send(d []byte) error
+}
+
+// Streamer is an [io.ReadWriter] that transfers data over
+// a [Stream]. All reads consume data from the stream and
+// all writes send data over the stream.
+type Streamer struct {
+	wLock  sync.Mutex
+	stream Stream
+
+	rLock  sync.Mutex
+	buffer []byte
+}
+
+// NewStreamer constructs a Streamer that operates on the provided
+// Stream.
+func NewStreamer(stream Stream) (*Streamer, error) {
+	if stream == nil {
+		return nil, trace.BadParameter("parameter stream required")
+	}
+
+	return &Streamer{
+		stream: stream,
+	}, nil
+}
+
+// Read consumes data from the [Stream]. If a read doesn't consume
+// the full payload of a frame the remaining payload is buffered and
+// will be returned on subsequent calls prior to receiving more data
+// from the [Stream].
+func (s *Streamer) Read(p []byte) (int, error) {
+	s.rLock.Lock()
+	defer s.rLock.Unlock()
+
+	// return any buffered data prior to
+	// pulling from the stream again
```

Comment:
nit: the comment is kind of describing the inverse of what the code is doing.

maybe rephrase it?

```
// read from the stream only if we don't have 
// data leftover from previous calls
```